dashboard/lib/dynamic_config/dcdo.rb

```diff
@@ -0,0 +1,57 @@
+# A dynamic configuration module that allows us to update
+# config settings without pushing new code.
+
+require 'dynamic_config/datastore_cache'
+require 'dynamic_config/adapters/dynamodb_adapter'
+require 'dynamic_config/adapters/json_file_adapter'
+require 'dynamic_config/adapters/memory_adapter'
+
+class DCDOBase
+  def initialize(datastore_cache)
+    @datastore_cache = datastore_cache
+  end
+
+  # @param key [String]
+  # @param default
+  # @returns the stored value at key
+  def get(key, default)
+    raise ArgumentError unless key.is_a? String
+    value = @datastore_cache.get(key)
+
+    if !value.nil?
+      return value
+    end
+
+    default
+  end
+
+  # Sets the value for a given key
+  # @param key [String]
+  # @param value [JSONable]
+  def set(key, value)
+    raise ArgumentError unless key.is_a? String
+    @datastore_cache.set(key, value)
+  end
+
+  # Clear all stored settings
+  def clear
+    @datastore_cache.clear
+  end
+
+  # Factory method for creating DCDOBase objects
+  # @returns [DCDOBase]
+  def self.create
+    if Rails.env.test?
+      adapter = MemoryAdapter.new
+    elsif CDO.use_dynamo_tables
+      adapter = DynamoDBAdapter.new CDO.gatekeeper_table_name
+    else
+      adapter = JSONFileDatastoreAdapter.new CDO.gatekeeper_table_name
+    end
+
+    datastore_cache = DatastoreCache.new adapter
+    DCDOBase.new datastore_cache
+  end
+end
+
+DCDO = DCDOBase.create
```

Comment:
This will result in an uncaught exception if DynamoDB is unavailable, which will prevent the web server from starting.   That seems bad because it could prevent us from starting any web servers even though this DynamoDB isn't critical to serving the HOC levels.   It seems like we would want to catch failure, and use a fail-safe configuration until we are eventually able to refresh the configuration using dynamodb.  Perhaps the "fail-safe" configuration returns null for all keys, and it is up to the application decide what "fail-safe" means for each of the dynamic config values.